synapse/lib/oauth.py

```diff
@@ -0,0 +1,398 @@
+import copy
+import heapq
+import asyncio
+import logging
+
+import aiohttp
+
+import synapse.exc as s_exc
+import synapse.common as s_common
+
+import synapse.lib.coro as s_coro
+import synapse.lib.nexus as s_nexus
+import synapse.lib.config as s_config
+import synapse.lib.lmdbslab as s_lmdbslab
+
+logger = logging.getLogger(__name__)
+
+reqValidProvider = s_config.getJsValidator({
+    'type': 'object',
+    'properties': {
+        'iden': {'type': 'string', 'pattern': s_config.re_iden},
+        'name': {'type': 'string'},
+        'flow_type': {'type': 'string', 'default': 'authorization_code', 'enum': ['authorization_code']},
+        'auth_scheme': {'type': 'string', 'default': 'basic', 'enum': ['basic']},
+        'client_id': {'type': 'string'},
+        'client_secret': {'type': 'string'},
+        'scope': {'type': 'string'},
+        'auth_uri': {'type': 'string'},
+        'token_uri': {'type': 'string'},
+        'redirect_uri': {'type': 'string'},
+        'extensions': {
+            'type': 'object',
+            'properties': {
+                'pkce': {'type': 'boolean'},
+            },
+            'additionalProperties': False,
+        },
+        'extra_auth_params': {
+            'type': 'object',
+            'additionalProperties': {'type': 'string'},
+        },
+    },
+    'additionalProperties': False,
+    'required': ['iden', 'name', 'client_id', 'client_secret', 'scope', 'auth_uri', 'token_uri', 'redirect_uri'],
+})
+
+reqValidTokenResponse = s_config.getJsValidator({
+    'type': 'object',
+    'properties': {
+        'access_token': {'type': 'string'},
+        'expires_in': {'type': 'number', 'exclusiveMinimum': 0},
+    },
+    'additionalProperties': True,
+    'required': ['access_token', 'expires_in'],
+})
+
+class OAuthManager(s_nexus.Pusher):
+    '''
+    Organize and execute OAuth token refreshes.
+    '''
+
+    async def __anit__(self, cell, nexsroot=None):
+        await s_nexus.Pusher.__anit__(self, 'oauth', nexsroot=nexsroot)
```

Comment:
Do we have situations where the OAuthManager class uses a nexsroot that differs from the cell that owns it?

Reply:
Not currently; updated to just use cell.nexsroot